module/spock/spock_verifier.go

```diff
@@ -0,0 +1,191 @@
+package spock
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/onflow/flow-go/crypto"
+	"github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/state/protocol"
+)
+
+// Verifier provides functionality to verify SPoCK proofs
+type Verifier struct {
+	sync.Mutex
+
+	// state is used to query identities at a blockId to get StakingPublicKey
+	protocolState protocol.ReadOnlyState
+
+	// map of receipts by result ID that do not have matching SPoCK proofs
+	// For instance, if there are 5 receipts that have 5 SpockSets, say, Spockset1, Spockset2, Spockset3, Spockset4, and Spockset5.
+	// If Spockset1 and Spockset2 are for ER1, and they match with each other;
+	// And Spockset3, Spockset4, Spockset5 are for ER2. and they don't match with each other.
+	// Then there are 2 category, the first category has bucket: ER1_Spockset1, and the second category has buckets: ER2_Spockset3, ER2_Spockset4, and ER2_Spockset5.
+	// When we receive each approval, we will first check which category it should go, and find all the buckets to try matching against.
+	receipts map[flow.Identifier][]*flow.ExecutionReceipt
+
+	// map to cache and keep track of matched approvals
+	// approvals[resultID][approvalID] = receipt
+	approvals map[flow.Identifier]map[flow.Identifier]*flow.ExecutionReceipt
+}
+
+// NewVerifier creates a new spock verifier
+func NewVerifier(state protocol.ReadOnlyState) *Verifier {
```

Comment:
Suggest naming the implementation the same (or very similar) to the interface. In this case, we already have a verifier engine in the verification node. So I think naming it `Verifier` may not be a good idea. Maybe `SpockVerifier` is better.